File: blender_extension/roblox_interface/roblox_server/send_messages.py
from .pause_thread import PauseThread;

def register(utils):
    process_formats = utils.import_module("process_formats")

    class SendMessagesThread(PauseThread):
        def send_message(self, message_name, *args):
            assert message_name in process_formats.send_message_ids
            message_id = process_formats.send_message_ids[message_name]
            self.args_queued.append((message_id,) + args)
            self.unpause()
        
        def write_buffer(self, data, data_size):
            self.buf_len += data_size
            if self.buf_len <= process_formats.SEND_LIMIT:
                self.buffer.append(data)
            else:
                overflow = self.buf_len - process_formats.SEND_LIMIT
                self.buffer.append(data[0:overflow])
                self.buffers_queued.append(b"".join(self.buffer))
                self.buffer = [data[overflow:]]
                self.buf_len = overflow
                
        def parse(self, args, args_count, format_data, **masks):
            parsed_type = type(format_data)
            if parsed_type is list:
                for sub_format_data in format_data:
                    args_count = self.parse(args, args_count, sub_format_data, **masks)
            else:
                args_count = process_formats.write_funcs[format_data["type"]](self, args, args_count, format_data, **masks)

        def run(self):
            self.buffer = []
            self.buf_len = 0
            self.args_queued = []
            self.buffers_queued = []

            self.stop_thread = False
            while not self.stop_thread:
                if len(self.args_queued) == 0:
                    self.pause()
                args = self.args_queued.pop(0)
                message_id = args[0]
                self.write_buffer(process_formats.message_id_format.pack(message_id), process_formats.message_id_format.size)
                self.parse(args, 1, process_formats.message_formats[message_id])
                self.buffers_queued.append(b"".join(self.buffer))
                self.buffer = []
                self.buf_len = 0

    sendThread = SendMessagesThread(name = "blender_roblox_sync Send Messages Thread")
    global send_message
    def send_message(message_name, *args):
        sendThread.send_message(message_name, *args)

    global fetch_send_buffer
    def fetch_send_buffer():
        queue_len = len(sendThread.buffers_queued)
        if queue_len > 0:
            return sendThread.buffers_queued.pop(0)

    return {
        "threads": (sendThread,)
    }

File: roblox_plugin/src/Server/sendMessages.ts
import * as ProcessFormats from "./processFormats";
import PauseThread from "./pauseThread";

export interface SendMessagesThread extends PauseThread {
	buffer: buffer[];
	bufLen: number;
	buffersQueued: buffer[];
	argsQueued: defined[][];
}
export namespace SendMessagesThread {
	export function sendMessage(sendThread: SendMessagesThread, messageName: string, ...args: defined[]) {
		assert(messageName in ProcessFormats.sendMessageIds);
		const messageId = ProcessFormats.sendMessageIds.get(messageName);
		args.unshift(messageId as defined);
		sendThread.argsQueued.push(args);
		PauseThread.unpause(sendThread);
	}
	export function writeBuffer(sendThread: SendMessagesThread, data: buffer, dataSize: number) {
		sendThread.bufLen += dataSize;
		if (sendThread.bufLen <= ProcessFormats.SEND_LIMIT) {
			sendThread.buffer.push(data);
		} else {
			const overflow = sendThread.bufLen - ProcessFormats.SEND_LIMIT;
			const joinedBuffer = buffer.create(ProcessFormats.SEND_LIMIT);
			let offset = 0;
			for (const segment of sendThread.buffer) {
				buffer.copy(joinedBuffer, offset, segment);
				offset += buffer.len(segment);
			}
			buffer.copy(joinedBuffer, offset, data, 0, dataSize - overflow);
			sendThread.buffersQueued.push(joinedBuffer);
			sendThread.buffer.clear();

			const overflowSegment = buffer.create(overflow);
			buffer.copy(overflowSegment, 0, data, dataSize - overflow);
			sendThread.buffer.push(overflowSegment);
			sendThread.bufLen = overflow;
		}
	}

	export function parse(
		sendThread: SendMessagesThread,
		args: defined[],
		argsCount: number,
		formatData: ProcessFormats.FormatData,
		masks: Map<string, boolean>,
	): number {
		if (!("type" in formatData)) {
			for (const subFormatData of formatData) {
				argsCount = parse(sendThread, args, argsCount, subFormatData, masks);
			}
		} else {
			argsCount = ProcessFormats.writeFuncs.get(formatData.type)!(sendThread, args, argsCount, formatData, masks);
		}
		return argsCount;
	}

	export function create(): SendMessagesThread {
		const sendThread: SendMessagesThread = {
			buffer: [],
			bufLen: 0,
			buffersQueued: [],
			argsQueued: [],
			stopThread: false,
		};

		task.spawn(() => {
			while (!sendThread.stopThread) {
				if (sendThread.argsQueued.size() === 0) {
					PauseThread.pause(sendThread);
				}
				const args = sendThread.argsQueued.shift()!;
				const messageId = args[0] as number;
				const format = ProcessFormats.messageIdFormat;
				const data = buffer.create(format.size);
				format.write(data, 0, messageId);
				writeBuffer(sendThread, data, format.size);
				const masks = new Map<string, boolean>();
				parse(sendThread, args, 1, ProcessFormats.messageFormats[messageId], masks);

				const joinedBuffer = buffer.create(sendThread.bufLen);
				let offset = 0;
				for (const segment of sendThread.buffer) {
					buffer.copy(joinedBuffer, offset, segment);
					offset += buffer.len(segment);
				}
				sendThread.buffersQueued.push(joinedBuffer);
				sendThread.buffer.clear();
				sendThread.bufLen = 0;
			}
		});
		return sendThread;
	}
}
export default SendMessagesThread;
